runtime/nvqir/cutensornet/simulator_cutensornet.cpp

```diff
@@ -0,0 +1,373 @@
+    gate_tensor[offset] = {1.0, 0.0};
+    offset += (full_dim + 1);
+  }
+  // Set the target gate matrix:
+  for (std::size_t row = 0; row < target_gate_dim; ++row) {
+    for (std::size_t col = 0; col < target_gate_dim; ++col) {
+      const auto org_idx = row * target_gate_dim + col;
+      // The anchor point of the gate matrix inside the expanded matrix (lower
+      // right, i.e., shift up and left by target_gate_dim)
+      const auto block_anchor = full_dim - target_gate_dim;
+      // Row and column idxs in the expanded matrix
+      const auto block_row_idx = block_anchor + row;
+      const auto block_col_idx = block_anchor + col;
+      const auto expanded_idx = block_row_idx * full_dim + block_col_idx;
+      gate_tensor[expanded_idx] = target_gate[org_idx];
+    }
+  }
+  return gate_tensor;
+}
+
+void SimulatorTensorNetBase::applyGate(const GateApplicationTask &task) {
+  const auto &controls = task.controls;
+  const auto &targets = task.targets;
+  // Cache name lookup key:
+  // <GateName>_<num control>_<Param>
+  const std::string gateKey =
+      task.operationName + "_" + std::to_string(controls.size()) + "_" + [&]() {
+        std::stringstream paramsSs;
+        for (const auto &param : task.parameters) {
+          paramsSs << param << "_";
+        }
+        return paramsSs.str();
+      }();
+  const auto iter = m_gateDeviceMemCache.find(gateKey);
+
+  // This is the first time we see this gate, allocate device mem and cache it.
+  if (iter == m_gateDeviceMemCache.end()) {
+    void *dMem = allocateGateMatrix(
+        generateFullGateTensor(controls.size(), task.matrix));
+    m_gateDeviceMemCache[gateKey] = dMem;
+  }
+  std::vector<int32_t> qubits;
+  for (const auto &qId : controls)
+    qubits.emplace_back(qId);
+  for (const auto &qId : targets)
+    qubits.emplace_back(qId);
+  m_state->applyGate(qubits, m_gateDeviceMemCache[gateKey]);
+}
+
+void SimulatorTensorNetBase::applyExpPauli(
+    double theta, const std::vector<std::size_t> &controls,
+    const std::vector<std::size_t> &qubitIds, const cudaq::spin_op &op) {
+  flushGateQueue();
+  cudaq::info(" [{} decomposing] exp_pauli({}, {})", name(), theta,
+              op.to_string(false));
+  std::vector<std::size_t> qubitSupport;
+  std::vector<std::function<void(bool)>> basisChange;
+  op.for_each_pauli([&](cudaq::pauli type, std::size_t qubitIdx) {
+    if (type != cudaq::pauli::I)
+      qubitSupport.push_back(qubitIds[qubitIdx]);
+
+    if (type == cudaq::pauli::Y)
+      basisChange.emplace_back([&, qubitIdx](bool reverse) {
+        rx(!reverse ? M_PI_2 : -M_PI_2, qubitIds[qubitIdx]);
+      });
+    else if (type == cudaq::pauli::X)
+      basisChange.emplace_back([&, qubitIdx](bool) { h(qubitIds[qubitIdx]); });
+  });
+
+  if (!basisChange.empty())
+    for (auto &basis : basisChange)
+      basis(false);
+
+  std::vector<std::pair<std::size_t, std::size_t>> toReverse;
+  for (std::size_t i = 0; i < qubitSupport.size() - 1; i++) {
```

Comment:
If the Pauli type has all `I`'s in it, `qubitSupport` will be empty, and the `- 1` will wrap around. I imagine we should probably add a check for that condition.

Reply:
I've added a check for this condition here as well as in qpp.